feat: add autoware_speed_scale_corrector package

yhisaki opened the pull request:
## Description

Documentation is [here](https://autowarefoundation.github.io/autoware_universe/pr-11347/sensing/autoware_speed_scale_corrector/)!

This PR introduces a new package `autoware_speed_scale_corrector` that performs vehicle speed scale correction. The package estimates and corrects speed scale factors by comparing velocities calculated from odometry with the VelocityReport from the vehicleto address discrepancies.

### Algorithm
1. **Data Collection and Buffering**: Collection of pose, IMU, and velocity report data
2. **Time Synchronization**: Identification of common time intervals across all sensor data streams
3. **Data Preprocessing**: Noise reduction using Gaussian smoothing (σ=0.7)
4. **Trajectory Interpolation**: Cubic spline interpolation (position) and linear interpolation (angular velocity, velocity)
5. **State Vector Creation**: Calculation of odometry velocity, distance, integrated distance, and angular velocity
6. **Constraint Validation**: Constraint checks for angular velocity, speed, and speed change
7. **Scale Factor Estimation**: Calculation of instantaneous scale factor and moving average update

### Inputs/Outputs
**Inputs**:
- `~/input/pose_with_covariance`: Odometry information
- `~/input/velocity_report`: Vehicle velocity report
- `~/input/imu`: IMU sensor data

**Outputs**:
- `~/output/scale_factor`: Estimated speed scale factor
- `~/output/debug_info`: Debug information

### Parameters
- `time_window`: Time window width for estimation (default: 4.0s)
- `time_interval`: Data sampling interval (default: 0.1s)
- `initial_speed_scale_factor`: Initial scale factor (default: 1.0)
- `max_angular_velocity`: Maximum angular velocity constraint (default: 1.0 rad/s)
- `max_speed`/`min_speed`: Speed constraints (default: 15.0/2.0 m/s)
- `max_speed_change`: Maximum speed change constraint (default: 1.0 m/s²)

## Related links

**Parent Issue:**

- Link

<!-- ⬇️🟢
**Private Links:**

- [CompanyName internal link]()
⬆️🟢 -->

## How was this PR tested?

Report:
https://tier4.atlassian.net/wiki/x/mgHA_g

## Notes for reviewers

None.

## Interface changes

None.

<!-- ⬇️🔴

### Topic changes

#### Additions and removals

| Change type   | Topic Type      | Topic Name    | Message Type        | Description       |
|:--------------|:----------------|:--------------|:--------------------|:------------------|
| Added/Removed | Pub/Sub/Srv/Cli | `/topic_name` | `std_msgs/String`   | Topic description |

#### Modifications

| Version | Topic Type      | Topic Name        | Message Type        | Description       |
|:--------|:----------------|:------------------|:--------------------|:------------------|
| Old     | Pub/Sub/Srv/Cli | `/old_topic_name` | `sensor_msgs/Image` | Topic description |
| New     | Pub/Sub/Srv/Cli | `/new_topic_name` | `sensor_msgs/Image` | Topic description |

### ROS Parameter Changes

#### Additions and removals

| Change type   | Parameter Name | Type     | Default Value | Description       |
|:--------------|:---------------|:---------|:--------------|:------------------|
| Added/Removed | `param_name`   | `double` | `1.0`         | Param description |

#### Modifications

| Version | Parameter Name   | Type     | Default Value | Description       |
|:--------|:-----------------|:---------|:--------------|:------------------|
| Old     | `old_param_name` | `double` | `1.0`         | Param description |
| New     | `new_param_name` | `double` | `1.0`         | Param description |

🔴⬆️ -->

## Effects on system behavior

None.

YamatoAndo commented on sensing/autoware_speed_scale_corrector/src/node.cpp:
```diff
@@ -0,0 +1,105 @@
+SpeedScaleEstimatorParameters load_parameters(rclcpp::Node * node)
+{
+  SpeedScaleEstimatorParameters parameters;
+  parameters.time_window = node->declare_parameter<double>("time_window", 4.0);
+  parameters.time_interval = node->declare_parameter<double>("time_interval", 0.1);
+  parameters.initial_speed_scale_factor =
+    node->declare_parameter<double>("initial_speed_scale_factor", 1.0);
+  parameters.max_angular_velocity = node->declare_parameter<double>("max_angular_velocity", 1.0);
+  parameters.max_speed = node->declare_parameter<double>("max_speed", 15.0);
+  parameters.min_speed = node->declare_parameter<double>("min_speed", 2.0);
+  parameters.max_speed_change = node->declare_parameter<double>("max_speed_change", 1.0);
```
Do not set a default value
https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/ros-nodes/parameters/#workflow

yhisaki replied:
Thank you!
The fix is in commit 4dde3cd84e56f16e3c9ad203480008e452fb2271.

YamatoAndo commented on sensing/autoware_speed_scale_corrector/include/autoware/speed_scale_corrector/speed_scale_estimator.hpp:
```diff
@@ -0,0 +1,137 @@
+ */
+struct SpeedScaleEstimatorParameters
+{
+  double time_window = 4.0;                 //!< Time window width used for estimation [s]
+  double time_interval = 0.1;               //!< Data sampling interval [s]
+  double initial_speed_scale_factor = 1.0;  //!< Initial scale factor
+  double max_angular_velocity = 1.0;        //!< Maximum angular velocity constraint [rad/s]
+  double max_speed = 15.0;                  //!< Maximum speed constraint [m/s]
+  double min_speed = 2.0;                   //!< Minimum speed constraint [m/s]
+  double max_speed_change = 1.0;            //!< Maximum speed change constraint [m/s²]
```
Do not set a default value

YamatoAndo commented on sensing/autoware_speed_scale_corrector/include/autoware/speed_scale_corrector/node.hpp:
```diff
@@ -0,0 +1,116 @@
+// Copyright 2025 TIER IV, Inc.
```
Are these header files referenced externally?
If not, move them under src.
https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/ros-nodes/directory-structure/#entire-structure

YamatoAndo commented on sensing/autoware_speed_scale_corrector/package.xml:
```diff
@@ -0,0 +1,35 @@
+  <name>autoware_speed_scale_corrector</name>
+  <version>0.3.0</version>
+  <description>The autoware_speed_scale_corrector package</description>
+  <author email="yukinari.hisaki.2@tier4.jp">Yukinari Hisaki</author>
+  <maintainer email="yukinari.hisaki.2@tier4.jp">Yukinari Hisaki</maintainer>
+  <maintainer email="takayuki.murooka@tier4.jp">Takayuki Murooka</maintainer>
+  <maintainer email="yamato.ando@tier4.jp">Yamato Ando</maintainer>
+
+  <license>Apache License 2.0</license>
```
```suggestion
  <maintainer email="yukinari.hisaki.2@tier4.jp">Yukinari Hisaki</maintainer>
  <maintainer email="takayuki.murooka@tier4.jp">Takayuki Murooka</maintainer>
  <maintainer email="yamato.ando@tier4.jp">Yamato Ando</maintainer>

  <license>Apache License 2.0</license>
  
  <author email="yukinari.hisaki.2@tier4.jp">Yukinari Hisaki</author>

```

YamatoAndo replied:
> File 'package.xml' is invalid:
> 5: /__w/autoware_universe/autoware_universe/sensing/autoware_speed_scale_corrector/package.xml:7: element author: Schemas validity error : Element 'author': This element is not expected. Expected is ( maintainer ).

yhisaki replied:
Thank you!
The fix is in commit c9e9993681417102db1f95ccc9505abf00a8c7fe.

yhisaki committed:
fix: correct spelling and variable naming in speed scale corrector node and utilities

- Updated documentation to use "ROS 2" instead of "ROS2" for consistency.
- Renamed variables in speed scale estimator and utility functions for clarity, changing `dxdt` to `dx_dt` and `wsum` to `w_sum`.

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
fix package.xml

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
refactor: update CMake configuration and launch file for speed scale corrector

- Updated CMake minimum version requirement from 3.5 to 3.14.
- Refactored library and executable definitions to use the project name variable for consistency.
- Removed the main.cpp file and transitioned to using rclcpp_components for node registration.
- Updated launch file to reflect the new executable name for the speed scale corrector node.

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
chore: add author tag back to package.xml for autoware_speed_scale_corrector

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
refactor: make SpeedScaleCorrectorNode header private

- Moved node.hpp from include/ to src/ directory to hide implementation details from external access
- Updated include statements in main.cpp and node.cpp to use local header path
- This change encapsulates the node implementation and prevents external dependencies on internal structures

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
refactor: update SpeedScaleEstimatorParameters to use default initialization

- Changed member variables in SpeedScaleEstimatorParameters to use default initialization instead of hardcoded values.
- Updated the load_parameters function to declare parameters without default values, allowing for runtime configuration.

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
feat: add autoware_speed_scale_corrector package

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
feat: add JSON schema for speed scale corrector parameters

- Introduced a new JSON schema file to define and validate parameters for the speed scale corrector module.
- Updated README to reference the new schema instead of listing parameters directly.

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki commented on sensing/autoware_speed_scale_corrector/README.md:
```diff
@@ -0,0 +1,87 @@
+| `max_angular_velocity`       | double | 1.0     | Maximum angular velocity constraint [rad/s] |
+| `max_speed`                  | double | 15.0    | Maximum speed constraint [m/s]              |
+| `min_speed`                  | double | 2.0     | Minimum speed constraint [m/s]              |
+| `max_speed_change`           | double | 1.0     | Maximum speed change constraint [m/s^2]     |
```
fixed in 91963b747d9509c29cd673bff870fb55bba851bd.

yhisaki committed:
docs: enhance README formatting for speed scale corrector

- Improved the formatting of mathematical equations in the README for better readability.
- Added line breaks and spacing around equations to enhance clarity and presentation.

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
refactor: remove use_sim_time argument from speed scale corrector launch file

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
refactor: update output argument paths in speed scale corrector launch file

- Changed default values for output_scale_factor and output_debug_info arguments to include the correct namespace for better organization.

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
refactor: update speed scale corrector parameters and node implementation

- Replaced time_window and time_interval parameters with update_hz in speed_scale_corrector.yaml.
- Modified SpeedScaleEstimatorParameters to include covariance parameters for initial speed scale factor, process noise, and measurement noise.
- Updated the SpeedScaleEstimator to use a vector of PoseStamped instead of PoseWithCovarianceStamped for pose input.
- Refactored node implementation to accommodate changes in parameter handling and pose processing.
- Enhanced the update method to utilize Kalman filter for speed scale estimation.

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
refactor: update input pose argument in speed scale corrector launch file

- Changed the default input pose argument from pose_with_covariance to pose for improved clarity and consistency in data handling.

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
refactor: update speed scale corrector parameters and implementation

- Changed parameter name from update_hz to update_interval in speed_scale_corrector.yaml and related code.
- Enhanced SpeedScaleEstimator to include additional state variables for improved estimation.
- Introduced a new utility function to calculate time differences between poses.
- Updated node implementation to reflect changes in parameter handling and state updates.

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
feat: add get_update_interval method to SpeedScaleEstimator

- Introduced a new method to retrieve the update interval for the speed scale estimator.
- Updated SpeedScaleCorrectorNode to utilize the new method for timer configuration, enhancing parameter handling.

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
fix: update previous pose in speed scale estimator

- Added assignment of the current pose to previous_pose_ when the time difference exceeds twice the update interval, ensuring accurate pose tracking for speed scale estimation.

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
refactor: simplify include paths in speed scale corrector

- Updated include statements in node.cpp, node.hpp, speed_scale_estimator.cpp, and utils.cpp to remove the autoware namespace prefix, improving code readability and maintainability.
- Introduced new header files for speed_scale_estimator and utils, encapsulating related functionalities.

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>
